packages/utilities/src/asAsync.tsx

```diff
@@ -0,0 +1,166 @@
+import * as React from 'react';
+
+interface IAsAsyncState<TProps> {
+  component?: React.ComponentType<TProps> | undefined;
+  error?: Error;
+  placeholder?: JSX.Element;
+  onLoad?: () => void;
+  onError?: (error: Error) => void;
+}
+
+/**
+ * If possible, use a WeakMap to maintain a cache of loaded components.
+ * This can be used to synchronously render components that have already been loaded,
+ * rather than having to wait for at least one async tick.
+ */
+const syncModuleCache =
+  typeof WeakMap !== 'undefined'
+    ? new WeakMap<() => Promise<any>, React.ComponentType<any>>() // tslint:disable-line:no-any
+    : undefined;
```

Comment:
if weakmap does not exist, this ends up being undefined, which will cause nullrefs later...